### PR TITLE
Document Manager

### PR DESCRIPTION
**Pull request by dantleech**

### CODE REVIEW

**Comment by danrot** [comment]
I am just adding my thoughts as a base for discussion into a tickable list here, since this PR is too big to add the comments on the single files. Once we have discussed and fixed the issues or decided on my doubts are invalid the items can be ticked off.

- [ ] As you already know I would suggest to give the document manager a certain responsibility rather than just let it dispatch events, which should IMHO only be used to extend this functionality from the outside. The way it is currently implemented is really hard to understand what's happening. Functionality could still be split using multiple services, which then are used by the `DocumentManager` (Then it would make even more sense to add the document-manager to sulu/sulu than it already does)
- [ ] I also don't like the fact that some events dispatch other events (e.g. the `FindSubscriber` dispatches the `HydrateEvent`, which could easily lead to endless loops
- [ ] Using events for stuff like query create and configure options also feels very strange, since there are other mechanisms to do stuff like this (configure options sounds like this should be done using the container)
- [ ] IMO using reflections to access non-public properties is an absolute no-go. E.g. the `NodeNameBehavior` should have not only a `getNodeName` but also a `setNodeName` in order for the system to access these properties. This way the it doesn't matter what internal structure the Document has, and just a Note in the interface saying there has to be a property added isn't a very good hint and is not what the developer expects. If there is a private variable it should really only be updated by this class, everything else can lead to errors which are hard to find.
- [ ] A representation like `S\C\C\D\S\ExtensionSubscriber` in the logs doesn't really improve readability IMHO, and makes distinguishing very hard if you have multiple classes called `ExtensionSubscriber`
- [ ] The advantage of easy profiling you have mentioned once could easily be achieved using the `Stopwatch` without the overhead of event dispatching.
- [ ] I could imagine that having own listener for every single property like UUID  including an own class instantiation and event dispatching might lead to quite a big overhead. E.g. in doctrine the hydration is also very heavy, I'd prefer not to make it even worse with loads of events.

Summed up, I really like the fact that the `ContentMapper` already got a lot slimmer and the responsibility is much clearer divided over multiple classes, but I think this could also be achieved without a such extensive event usage, whose disadvantages IMHO outweigh the advantages.

Sry about the big comment, but as I said it is really hard to review this PR...

**Comment by dantleech** [comment]
>>  IMO using reflections to access non-public properties is an absolute no-go.

The where no setter is provided, then the property is not supposed to be set by the user. For example, `node name` is set automatically with the AutoName behavior. The webspace name is another example. You cannot set the webspace name.

>>  I could imagine that having own listener for every single property like UUID

We can optimize this and remove not-a-few listeners by introducing a mapping listener which simply maps PHPCR properties to properties from a configured map (similar to standard ORM mappings). This would be a necessary future feature.

>> events

In any implementation of this, events would play a central part. The PHPCR-ODM version had a similar number of events. Building all of this functionality into <some class> wouldn't help, and I honestly don't see an alternative. The only difference here is that the core architecture is built on events. I believe any alternative strategy would strongly resemble the event pattern.

For your other comments, I do not share your misgivings. We can talk tommorow.

**Comment by chirimoya** [comment]
Current issues:
- [ ] Smart Content rendering doesn't work; nor in preview or on the website.
- [ ] Column navigation doesn't display last selected tree. Children doesn't get loaded. Only the first level.

File: src/Sulu/Component/Content/Compat/Property.php [hunk 1]
```diff
@@ -227,9 +221,14 @@ public function getMultilingual()
...
 
     /**
+<<<<<<< HEAD:src/Sulu/Component/Content/Property.php
      * returns tags defined in xml.
      *
      * @return \Sulu\Component\Content\PropertyTag[]
+=======
+     * returns tags defined in xml
+     * @return \Sulu\Component\Content\Compat\PropertyTag[]
+>>>>>>> Moving stuff:src/Sulu/Component/Content/Compat/Property.php
```
Comment (dantleech) [inline src/Sulu/Component/Content/Compat/Property.php]: Fix.

File: src/Sulu/Component/Content/Document/Subscriber/ExtensionSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        foreach ($extensions as $extension) {
+            $extensionData = null;
+
+            if (!isset($extensionsData[$extension->getName()])) {
+                continue;
+            }
+
+            $extensionData = $extensionsData[$extension->getName()];
+
+
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Subscriber/ExtensionSubscriber.php]: EL

**Comment by danrot** [comment]
I guess you won't be able to remove the `Compat` layer before you leave, right?

**Comment by dantleech** [comment]
To remove the `Compat` namespace we would mainly need to migrate: content types, preview, structure resolver (the view) and resource locator, as we discussed.

This will take months I expect.

**Pull request merged**